### PR TITLE
[MRG] HW7

### PR DESCRIPTION
**Pull request by kjacks21**
## Assignment

**Goal**: Polymorphic Abstraction.
Liskov 8.8 and 8.9. (page 205) You are welcome to use Bloch's suggested exception types instead of Liskov's. 
Use generics when defining your Indexer interface. That is, the search method should have generic parameters instead of Object parameters. Your code should compile without warnings. With generics, the signature for search() should be:

` public static <E> int search (Indexer<E> c, E x) `
Supply JUnit tests for your implementation.

In 8.9, instead of VectorIndexer, define and implement ListIndexer over Java List collections.

Java List collections typically allow null entries. Make sure you handle null values (either in c or for x) correctly.

## Tasks

- [ ] JUnit tests

- [ ] 8.8

- [ ] 8.9

### CODE REVIEW

**Comment by 1000DS000000** [comment]
It's confusing to me as well, but i think its something similar to the Adder interface on p199. I think we need to have 2 additional methods in the Indexer that will help the search method. One to see if it is Searchable, and the other to get the size. I'm not sure if those 2 methods need to be implemented though.  I think in 8.9 we implement it.

**Comment by 1000DS000000** [comment]
I'm not sure if this is correct, but i think we also need to define a get method to the Indexer interface to be used in the search method.

**Comment by kjacks21** [comment]
Sorry for the delay on this - been working on another project. Reading the chapter now and then I'll contribute to the discussion/code.

**Comment by ngreenwald89** [comment]
Yeah who wants what?

**Comment by kjacks21** [comment]
Do we want to get on a call together this evening and hash this out?

**Comment by 1000DS000000** [comment]
I'm not going to be able to. Got class till 10.20 at the latest.

**Comment by ngreenwald89** [comment]
Leaving office now let's talk when I get home.

File: hw7/src/ListIndexer.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+public class ListIndexer implements Indexer {
+    public static <E> int search (Indexer<E> c, E x) throws
+            NullPointerException, NotFoundException, ClassCastException {
```
Comment (kjacks21) [inline hw7/src/ListIndexer.java]: What should we throw here instead of `NotFoundException`?

**Comment by kjacks21** [comment]
@1000DS000000 Nat and I are pair programming right now and can do a google hangout when you become available. Since Nat and I can't work on this tomorrow and we are kind of lost, it'd be good if we can get this done tonight.

**Comment by 1000DS000000** [comment]
I'm somewhat available now, waiting to catch bus from campus. I'm not sure how Google hangouts work. also my phone is somewhat low on battery.

**Comment by kjacks21** [comment]
Google hangouts makes it so that we can screenshare, but it probably isn't necessary. Nat had to checkout for the night, but I'm going to be working on this tonight. When you get back home and are willing to stay up, we can either correspond over here or phone/google chat to finish this. To summarize some things from piazza that are relevant:

On the use of equals: https://piazza.com/class/jl84sbpvd70246?cid=53
On what searchable means: https://piazza.com/class/jl84sbpvd70246?cid=52
On where to implement the `search` method: https://piazza.com/class/jl84sbpvd70246?cid=54
- Here there appears to be debate as to whether `search` should be implemented fully in the interface and overridden in the concrete class, or if it should just be implemented in the concrete class.

**Comment by 1000DS000000** [comment]
I'll b able to, but its gonna b another hr till I get home.

**Comment by 1000DS000000** [comment]
what I'm thinking is that the indexer is like a helper class. it just takes in a list or some collection that needs to find the index of some element or the size of it.

**Comment by 1000DS000000** [comment]
just got home. I think I have an idea for the skeleton of the code.

**Comment by 1000DS000000** [comment]
i added my thoughts in comments in the Indexer class... i am not sure though if the parameters should b Collection or just Objects.

**Comment by kjacks21** [comment]
Go ahead and make the changes that you have in your comments since I have no objections to them.

**Comment by kjacks21** [comment]
I'm going to hit the hay and will try to look at this tomorrow as well. I also asked for an extension since GitHub having an outage starting Sunday and clearing up not to long ago today was disruptive to us.

**Comment by 1000DS000000** [comment]
another thing i think could be possible. instead of having the search method in its own class.. can't we have it in the JUnit test class. and ok.. i think i figured out the search.. will post what i have to GitHub.

**Comment by kjacks21** [comment]
Great! Thanks for working on this - I really appreciate it.

**Comment by kjacks21** [comment]
Laurel denied the extension, saying no exceptions. I thought that was just for quizzes? 😔

**Comment by 1000DS000000** [comment]
well i think i got it. all that is left is the jUnit.

**Comment by kjacks21** [comment]
I can do the tests. Do we still need `Searchable.java`?

**Comment by 1000DS000000** [comment]
i don't think so

**Comment by kjacks21** [comment]
Someone said that having `search` in the JUnit test isn't a good idea. Thoughts? Source: https://piazza.com/class/jl84sbpvd70246?cid=54

**Comment by 1000DS000000** [comment]
I'm not sure then how to go about implementing that then. I've been getting errors because of it being a static method. I'll see if i can move it back to the searchable class, but if not. i think we should go with what we have.

**Comment by 1000DS000000** [comment]
so I've found a way, but in the Searchable class, the list wont b a generic type, itll b a String

**Comment by kjacks21** [comment]
What it make sense to implement the it in the interface?

**Comment by 1000DS000000** [comment]
maybe in the ListIndexer

**Comment by 1000DS000000** [comment]
the problem is. is that the static method has to make the list to be searched static, in order for it to work. and if i make the list<E> static, the compiler it still telling me that it can't because E is not static.

**Comment by kjacks21** [comment]
Is it possible to have the method not be static in the interface, but overridden

**Comment by 1000DS000000** [comment]
i think it needs to have the same signature, even if we override it

**Comment by kjacks21** [comment]
are you on campus? I am if you want to link up

**Comment by 1000DS000000** [comment]
i am.. I'm on 2nd floor of jc by the library

**Comment by kjacks21** [comment]
omw

**Comment by kjacks21** [comment]
Assignment submitted. Thanks all.

**Comment by kjacks21** [comment]
Feedback from grader:

"Search requires a way to find elements of c by their index and to determine the size of c.".

So you need to define a method that takes an index as parameter, and return an element (not the other way around):

public interface Indexer<E> {

 /**

  * Return the element at the given index in this.

  * Index must be within range (i.e. 0 <= index < size()).

  *

  * @throws IndexOutOfBoundsException if the index is out of range.

  */

    E get(int index);

 /**

  * Return the size of this. If this is empty, return 0.

  */

    int size();

}

**Pull request merged**